[CUBRIDMAN-251] Phase1 PL/CSQL Language features

hyunikn opened the pull request:
http://jira.cubrid.org/browse/CUBRIDMAN-251

Phase 1의 PL/CSQL 언어 feature에 대한 한글 내용 업데이트 입니다. 
예제를 더 추가해야 하지만, 일단 내용 초안이 잡혀서 PR 올립니다. 
각각의 소주제 별로 별도 commit 으로 되어 있으니 변경 사항을 한꺼번에 보시는 것보다 
commit 별로 나눠서 보시는게 파악이 편하실 수 있습니다. 

렌더링된 문서 링크는 해당 지라 이슈의 댓글에서 보실 수 있습니다. 

------------
추가 (2024-12-10): 예제도 모두 추가했습니다.

hyunikn committed:
update reserved words table

hyunikn committed:
refine phrases related to Server Parameters

hyunikn committed:
. delete the phrased on Dymaic SQL For Loop
. minor refinements

hyunikn committed:
minor updates

hyunikn committed:
add a section on %ROWTYPE

hyunikn committed:
Merge branch 'feature/plcsql' into phase1-pl-features

hyunikn committed:
change URL links to internal links (:ref:`...`)

hyunikn committed:
unhandled exception gives position and message information to the client (CBRD-24802)

hyunikn committed:
syntax and semantic checks during the execution of CREATE PROCEDURE/FUNCTION statements (CBRD-24733)

hyunikn committed:
. sqlcode of system built-in exceptions and others (CBRD-24819)
. section of raise_application_error statement

hyunikn committed:
built-in function IF cannot be used

hyunikn committed:
concatenation's precedence is less than addition and subtraction (CBRD-25130)

hyunikn committed:
respect NUMERIC's precision and scale when they are specified by %TYPE in a function return type. (CBRD-25682)

hyunikn committed:
description of a CSQL session command ;server-output (CBRD-25415)

hyunikn committed:
how to ignore string literal, comment, identifier blocks when running session commands (CBRD-25633)

hyunikn committed:
more on sementic error cases

hyunikn committed:
more on runtime exception cases

hyunikn committed:
mark places to insert examples

hyunikn committed:
Merge branch 'feature/plcsql' into phase1-pl-features

hyunikn committed:
minor: 파라미터 --> 인자, 인자 타입 --> 인자타입, 리턴 타입 --> 리턴타입

kangmin5505 commented:
리뷰 대신 pr을 남겼습니다.
https://github.com/hyunikn/cubrid-manual/pull/1

swi0110 commented on ko/csql.rst:
```diff
@@ -341,7 +341,42 @@ CSQL 시작 옵션
+CSQL은 SQL의 문자열 리터럴, 주석, 식별자를 (참고: :ref:`작성 규칙 <lexical_rules>`) 인식하면서 동작한다. 특히,
+
+* 작은 따옴표로 둘러싸인 문자열 리터럴
+* '/\*'와 '\*/'로 둘러싸인 C 언어 방식 주석
```
주석의 경우 아래와 같이 shell script 방식으로도 처리 가능한줄 압니다.
확인 부탁드립니다.

```
csql> select -- comment 'aaa'                                                                                                                                                                                                                                                
csql> 'ddd' from dual;                                                                                                                                                                                                                                                             

=== <Result of SELECT Command in Line 1> ===

  'ddd'               
======================
  'ddd'
```

hyunikn replied:
-- 는 SQL 주석이군요.
shell script 방식이란 어떤 것을 말씀하시는 것인지요.

swi0110 replied:
제가 헷갈려서 sql 주석을 shell script 방식이라고 잘못 서술했습니다

hyunikn replied:
말씀하신 대로 모든 타입의 주석을 언급하는 것이 좋겠네요.
수정했습니다.

swi0110 commented on ko/csql.rst:
```diff
@@ -1364,6 +1403,41 @@ OPT LEVEL의 상세한 내용은 :ref:`viewing-query-plan`\ 를 참고한다.
+
+이 값을 ON으로 설정하면 서버의 DBMS_OUTPUT 버퍼에 저장된 메시지를 출력한다. 기본값은 OFF이다.
+DBMS_OUTPUT 버퍼는 주로 PL/CSQL 저장 프로시저/함수에서 DBMS_PUTPUT.put_line() 호출을 통해 쌓인 메시지들을 저장하고 있다.
+DBMS_OUTPUT 메시지는 CSQL이 실행한 SQL 문의 결과 출력에 뒤이어 '<DBMS_PUTPUT>' 이라는 표시 다음에 출력된다.
```
```suggestion
DBMS_OUTPUT 메시지는 CSQL이 실행한 SQL 문의 결과 출력에 뒤이어 '<DBMS_OUTPUT>' 이라는 표시 다음에 출력된다.
```

hyunikn replied:
오타입니다. 감사합니다.

swi0110 commented on ko/csql.rst:
```diff
@@ -1364,6 +1403,41 @@ OPT LEVEL의 상세한 내용은 :ref:`viewing-query-plan`\ 를 참고한다.
+이 값을 ON으로 설정하면 서버의 DBMS_OUTPUT 버퍼에 저장된 메시지를 출력한다. 기본값은 OFF이다.
+DBMS_OUTPUT 버퍼는 주로 PL/CSQL 저장 프로시저/함수에서 DBMS_PUTPUT.put_line() 호출을 통해 쌓인 메시지들을 저장하고 있다.
+DBMS_OUTPUT 메시지는 CSQL이 실행한 SQL 문의 결과 출력에 뒤이어 '<DBMS_PUTPUT>' 이라는 표시 다음에 출력된다.
+특히, 실행 순서 상 DBMS_PUTPUT.put_line() 호출이 먼저이고 에러 발생이 나중이라도 실행 결과인 에러메시지가 먼저 출력되고
```
특히, 실행 순서 상 DBMS_PUTPUT.put_line() 호출이 먼저이고 에러 발생이 나중이라도 실행 결과인 에러메시지가 먼저 출력되고
```suggestion
특히, 실행 순서 상 DBMS_OUTPUT.put_line() 호출이 먼저이고 에러 발생이 나중이라도 실행 결과인 에러메시지가 먼저 출력되고
```

swi0110 commented on ko/pl/pl.rst:
```diff
@@ -26,6 +28,16 @@ PL/CSQL은 저장 프로시저나 저장 함수를 생성하는데 사용된다.
+
+.. code-block:: sql
+
+    -- TODO: example
```
예제가 아니라 불필요한 내용이 추가된 것 같습니다.

hyunikn replied:
다음번 PR 에서 채울 예제 자리를 표시해 둔 것입니다.

swi0110 commented on ko/pl/pl.rst:
```diff
@@ -89,6 +101,25 @@ PL/CSQL은 저장 프로시저나 저장 함수를 생성하는데 사용된다.
             RETURN -1;
     END;
 
+CREATE PROCEDURE/FUNCTION 문을 실행하면 저장 프로시저/함수의 문법과 실행의미에 관련된 각종 규칙들을 검사한다.
```
문장이 다소 어색한 것 같아, 아래와 같이 제안을 드려봅니다.
```suggestion
CREATE PROCEDURE/FUNCTION 문을 실행하면 저장 프로시저/함수의 문법과 의미에 관한 각종 규칙들을 검사한다.
```

hyunikn replied:
실행의미는 Semantic을 번역한 것으로 일반적으로는 의미라고 할 수 있지만, 프로그래밍 용어로는 실행의미가 더 구체적인 표현 같습니다.

swi0110 commented on ko/pl/pl.rst:
```diff
@@ -107,15 +138,31 @@ SQL 구문 중에 다음에 해당하는 것들을 PL/CSQL 실행문으로 직
-SELECT 문의 INTO 절에 프로그램의 변수나 OUT 파라미터를 써서 조회 결과를 담을 수 있다.
+
+SELECT 문의 INTO 절에 프로그램의 변수나 OUT 인자를 써서 조회 결과를 담을 수 있다.
+이 때 조회 결과 값들의 갯수는 INTO 절 안의 변수나 OUT 인자의 갯수와 일치해야 하고
```
```suggestion
이 때 조회 결과 값들의 개수는 INTO 절 안의 변수나 OUT 인자의 개수와 일치해야 하고
```

hyunikn replied:
넵. 개수가 표준어네요. 감사합니다.

swi0110 commented on ko/pl/pl.rst:
```diff
@@ -107,15 +138,31 @@ SQL 구문 중에 다음에 해당하는 것들을 PL/CSQL 실행문으로 직
-SQL 구문의 문법과 의미는 CUBRID 매뉴얼 중
-`CUBRID SQL <https://www.cubrid.org/manual/ko/11.2/sql/index.html>`_\을 참고하도록 한다.
+SQL 구문의 문법과 의미는 CUBRID 매뉴얼 중 :ref:`CUBRID SQL <cubrid_sql>`\을 참고하도록 한다.
+다음은 Static SQL 사용예이다.
```
```suggestion
다음은 Static SQL 사용 예이다.
```

hyunikn replied:
네 수정하겠습니다. 감사합니다.

swi0110 commented on ko/pl/pl.rst:
```diff
@@ -265,31 +314,26 @@ Static/Dynamic SQL 밖의 PL/CSQL 문에서 아래 표의 단어들을 변수, 
 * SYS_REFCURSOR: 커서 변수를 선언할 때 사용한다.
   커서 변수의 용도는 :ref:`OPEN-FOR <cursor_manipulation>` 문을 참고한다.
-  BOOLEAN과 마찬가지로 CREATE PROCEDURE/FUNCTION 문에서 파라미터 타입이나 리턴 타입으로 SYS_REFCURSOR를 사용할 수 없고,
+  BOOLEAN과 마찬가지로 CREATE PROCEDURE/FUNCTION 문에서 인자타입이나 리턴타입으로 SYS_REFCURSOR를 사용할 수 없고
```
```suggestion
  BOOLEAN과 마찬가지로 CREATE PROCEDURE/FUNCTION 문에서 인자 타입이나 리턴 타입으로 SYS_REFCURSOR를 사용할 수 없고
```

hyunikn replied:
네. 수정하겠습니다. 감사합니다.

swi0110 commented on ko/pl/pl.rst:
```diff
@@ -459,19 +653,67 @@ PL/CSQL은 다른 많은 프로그래밍 언어와 마찬가지로 Exception 핸
 ==========================
 
-Static/Dynamic SQL 문의 동작은 각종 `서버 설정 <https://www.cubrid.org/manual/ko/11.2/admin/config.html#id2>`_\의 영향을 받는다.
+Static/Dynamic SQL 문의 동작은 :ref:`서버 설정 파라메터 <system_config>` 전체의 영향을 동일하게 받는다.
```
```suggestion
Static/Dynamic SQL 문의 동작은 :ref:`서버 설정 파라미터 <system_config>` 전체의 영향을 동일하게 받는다.
```

hyunikn replied:
수정하겠습니다. 감사합니다.

swi0110 commented on ko/pl/pl.rst:
```diff
@@ -953,15 +1242,37 @@ SYS_REFCURSOR 변수에 연결하고 그 SELECT 문의 결과를 조회해 오
+값들은 대응되는 변수나 OUT 인자에 대입 가능한 타입을 가져야 한다.
+
+SQL 문 실행 중에 에러가 나면 SQL_ERROR Exception이 발생한다.
+INTO 절을 포함한 경우 SELECT 문의 조회 결과는 한건 그리고 단 한건의 결과 레코드를 가져야 한다.
```
```suggestion
INTO 절을 포함한 경우 SELECT 문의 조회 결과는 한 건 그리고 단 한 건의 결과 레코드를 가져야 한다.
```

hyunikn replied:
네. 수정하겠습니다. 감사합니다.

swi0110 commented on ko/pl/pl.rst:
```diff
@@ -1112,7 +1430,7 @@ RETURN
 
 현재 루틴을 호출한 호출문 다음으로 분기한다.
-현재 루틴이 함수인 경우에는 그 함수의 리턴 타입에 맞는 반환값 *expression*\을 지정해야 한다.
+현재 루틴이 함수인 경우에는 그 함수의 리턴타입으로 변환 가능한 반환값 *expression*\을 지정해야 한다.
```
```suggestion
현재 루틴이 함수인 경우에는 그 함수의 리턴 타입으로 변환 가능한 반환값 *expression*\을 지정해야 한다.
```

hyunikn replied:
네. 수정하겠습니다. 감사합니다.

swi0110 commented on ko/pl/pl.rst:
```diff
@@ -1430,10 +1748,23 @@ FOR 문에서 SELECT 결과를 순회하기 위해 묵시적으로 선언되는
+    -- TODO: example
+
+호출되는 함수는 저장 함수, 내부 함수, 빌트인 함수 이렇게 세 가지 종류이다.
+이 중에서 빌트인 함수는 :ref:`연산자와 함수 <operators-and-functions>` 장에 나열된 큐브리드 내장 함수들을 말한다.
```
```suggestion
이 중에서 빌트인 함수는 :ref:`연산자와 함수 <operators-and-functions>` 장에 나열된 :ref:`큐브리드 내장 함수 <operators-and-functions>`들을 말한다.
```

hyunikn replied:
바로 위에 링크가 있어서 불필요 할 것 같습니다.

swi0110 commented on ko/pl/pl.rst:
```diff
@@ -1480,10 +1812,14 @@ CASE 표현식은 :ref:`CASE 실행문 <case_stmt>`\(Statement)과 마찬가지
-예외 처리 블럭 안에서 SQLCODE와 SQLERRM은 각각 현재 처리 중인 예외의 코드(INTEGER 타입)와 메시지(STRING 타입)를 나타낸다. 예외 처리 블럭 밖에서 SQLCODE와 SQLERRM은 각각 0과 'no error' 값을 갖는다.
+Exception 처리 블럭 안에서 SQLCODE와 SQLERRM은 각각 현재 처리 중인 Exception의 코드(INTEGER 타입)와
+에러메시지(STRING 타입)를 나타낸다.
+Exception 처리 블럭 밖에서 SQLCODE와 SQLERRM은 각각 0과 'no error' 값을 갖는다.
```
문서의 다른 부분에서는 '블럭'을 'block'이라고 영문 표기를 하고 있었습니다.
통일성을 위해 제안드립니다.

```suggestion
Exception 처리 block 밖에서 SQLCODE와 SQLERRM은 각각 0과 'no error' 값을 갖는다.
```

hyunikn replied:
차이 없으니 유지 해도 좋을 것 같습니다.

hyunikn replied:
수정하겠습니다. 감사합니다.

swi0110 commented on ko/pl/pl.rst:
```diff
@@ -1480,10 +1812,14 @@ CASE 표현식은 :ref:`CASE 실행문 <case_stmt>`\(Statement)과 마찬가지
 =================
 
-예외 처리 블럭 안에서 SQLCODE와 SQLERRM은 각각 현재 처리 중인 예외의 코드(INTEGER 타입)와 메시지(STRING 타입)를 나타낸다. 예외 처리 블럭 밖에서 SQLCODE와 SQLERRM은 각각 0과 'no error' 값을 갖는다.
+Exception 처리 블럭 안에서 SQLCODE와 SQLERRM은 각각 현재 처리 중인 Exception의 코드(INTEGER 타입)와
```
```suggestion
Exception 처리 block 안에서 SQLCODE와 SQLERRM은 각각 현재 처리 중인 Exception의 코드(INTEGER 타입)와
```

hyunikn replied:
수정하겠습니다. 감사합니다.

hyunikn committed:
fix typos

hyunikn committed:
fix more typos

hyunikn committed:
fix missing label

hyunikn committed:
describe all the kind of comments

hyunikn committed:
fix a single typo: Excption --> Exception

ctshim commented:
랜더링 된 페이지에서(http://192.168.2.208/ko/pl/pl.htm)
예제가 들어 갈 공간에 아래처럼 비워져 있는 부분은 이번에 다루지 않는 영역인가요?

-- TODO: example

hyunikn commented:
> 랜더링 된 페이지에서(http://192.168.2.208/ko/pl/pl.htm) 예제가 들어 갈 공간에 아래처럼 비워져 있는 부분은 이번에 다루지 않는 영역인가요?
> 
> -- TODO: example

금주 내로 별도 commit 으로 추가할 계획입니다. 
일단, 그 부분 빼고 검토 부탁드립니다.

ctshim commented on ko/pl/pl.rst:
```diff
@@ -357,30 +406,166 @@ ALTER PROCEDURE/FUNCTION <name> REBUILD 문을 실행해서 재컴파일 해주
+* s와 t의 필드 개수가 같다.
+* 각각의 필드 순번 i에 대해서, s와 t의 i번째 필드들의 타입을 S\ :sub:`i`\와 T\ :sub:`i`\라고 할 때, S\ :sub:`i`\에서 T\ :sub:`i`\로 대입 가능하다.
+
+레코드 변수 사이에 대입이 가능하기 위해서 같은 순번의 필드 이름들끼리 같을 필요는 없다.
```
타입이 같아야 한다 또는 묵시적인 cast가 가능해야 한다는 조건은 필요 없는 건가요?

hyunikn replied:
바로 윗줄의 "Si에서 Ti로 대입 가능하다" 가 그 뜻입니다.
"Si 와 Ti 가 같거나 Si에서 Ti 로 묵시적 형변환이 가능해야 한다" 가 좀 더 명확한가요?

ctshim replied:
좀 더 아래쪽에  다시 언급되는 내용이 있는데 여기서는 뭔가 표현이 되어 있네요

"Static/Dynamic SQL SELECT 문과 FETCH 문의 INTO 절에 레코드 변수를 쓸 수 있다. 단, 이 때 INTO 절 안에 다른 변수를 함께 쓸 수 없다. 그리고, 조회 결과를 이루는 컬럼들과 레코드 필드들의 개수가 같아야 한다. 이 때, 같은 순번의 컬럼과 필드끼리 이름이 같을 필요는 없지만 **컬럼 타입으로부터 필드 타입으로 대입 가능해야 한다**."

hyunikn replied:
네. 바로 윗 줄에도 "대입 가능" 하다 라는 표현이 있습니다.

ctshim commented on ko/pl/pl.rst:
```diff
@@ -953,15 +1243,37 @@ SYS_REFCURSOR 변수에 연결하고 그 SELECT 문의 결과를 조회해 오
+RAISE_APPLICATION_ERROR는 원하는 :ref:`코드와 에러메시지 <sqlcode>`\로 :ref:`Exception <exception>`\을
+일으키고자 할 때 사용한다.
+RAISE_APPLICATION_ERROR의 사용 형태는 Built-in 프로시저 호출처럼 보이지만 내부적으로는 PL/CSQL 실행문이다.
+첫번째 인자로 주는 코드는 1000보다 큰 INTEGER 값을 가져야 한다. 아니면, VALUE_ERROR Exception이 발생한다.
+1000 이하의 값은 시스템을 위해 예약되어 있기 때문이다.
```
여기서 말하는 시스템을 위해 예약되어 있는 값은 어디에 정의된 것인가요?

hyunikn replied:
사용자 정의 exception 번호가 시스템 빌트인과 겹치면 안될 것 같아서 제가 나름대로 정한 규약입니다. 
현재 구현이 그렇게 되어 있습니다.

bagus-kim commented on ko/csql.rst:
```diff
@@ -341,7 +341,42 @@ CSQL 시작 옵션
+* --, //, /\* \*/ 로 표시되는 주석
+* 큰따옴표, 대괄호, 백틱 부호로 둘러싸인 식별자
+
+안에 위치하는 문자열은 이들의 밖에서 가질 수 있는 고유의 의미를 잃고 오로지 문자열로만 인식되는데,
```
```suggestion
안에 위치하는 문자열은 이들의 밖에서 가질 수 있는 고유의 의미(세션 명령어로서의 기능)를 잃고 오로지 문자열로만 인식되는데,
```

hyunikn replied:
주어 '문자열은' 세션 명령어에 국한되지 않아서, 괄호 내용 추가는 맞지 않는 것 같습니다.

bagus-kim commented on ko/csql.rst:
```diff
@@ -341,7 +341,42 @@ CSQL 시작 옵션
+* 큰따옴표, 대괄호, 백틱 부호로 둘러싸인 식별자
+
+안에 위치하는 문자열은 이들의 밖에서 가질 수 있는 고유의 의미를 잃고 오로지 문자열로만 인식되는데,
+이것은 세션 명령어에 대해서도 마찬가지이다.
```
위에서 "고유의 의미를 잃고" 로 설명된 것이 아닌가요?

hyunikn replied:
네. 특히, 세션 명령어에 대해서도 그러하다는 뜻입니다.

bagus-kim commented on ko/csql.rst:
```diff
@@ -341,7 +341,42 @@ CSQL 시작 옵션
+
+안에 위치하는 문자열은 이들의 밖에서 가질 수 있는 고유의 의미를 잃고 오로지 문자열로만 인식되는데,
+이것은 세션 명령어에 대해서도 마찬가지이다.
+예를 들어, 아래 문자열 리터럴, 주석, 식별자 안의 ';exit'은 세션 명령어로서 동작하지 않는다. ::
```
```suggestion
예를 들어, 아래 문자열 리터럴, 주석, 식별자 안의 ';exit'은 세션 명령어로서 동작하지 않고, 문자열로 동작한다. ::
```

hyunikn replied:
네. 수정하겠습니다.

bagus-kim commented on ko/pl/pl.rst:
```diff
@@ -26,6 +28,16 @@ PL/CSQL은 저장 프로시저나 저장 함수를 생성하는데 사용된다.
+저장 프로시저/함수는 :ref:`큐브리드 내장 함수 <operators-and-functions>`\와 동일한 이름을 가질 수 없다.
+동일한 이름으로 선언하면 컴파일 과정에서 (CREATE 문 실행 과정에서) 에러가 발생한다.
+
+body 내부의 실행문들 중에 도달할 수 없는 것이 있을 때는 컴파일 과정에서 에러가 발생한다.
```
"도달할 수 없는 것이 있을 때는" 에 대한 설명이 추가해 보입니다.

hyunikn replied:
'실행 시간에' 라는 문구를 추가하면 조금 더 나아질 것 같습니다. 
그리고, 바로 아래에 예제를 추가할 계획입니다.

bagus-kim commented on ko/pl/pl.rst:
```diff
@@ -26,6 +28,16 @@ PL/CSQL은 저장 프로시저나 저장 함수를 생성하는데 사용된다.
+동일한 이름으로 선언하면 컴파일 과정에서 (CREATE 문 실행 과정에서) 에러가 발생한다.
+
+body 내부의 실행문들 중에 도달할 수 없는 것이 있을 때는 컴파일 과정에서 에러가 발생한다.
+다음은 도달할 수 없는 실행문이 있는 간단한 예이다.
```
"도달할 수 없는 실행문" 이란 무엇일까요? 이러한 용어가 사용되나요?

hyunikn replied:
아래 추가될 예제를 보시면 바로 이해가 되실 겁니다. 
예를 들어, return, raise 문 바로 다음에 있는 실행문이 있다면 그것은 도달할 수 없는 실행문이 됩니다.

hyunikn replied:
간단한 예제로 다음과 같은 것을 들 수 있겠습니다. 
```
    csql> CREATE OR REPLACE PROCEDURE test_unreachable_statement
    csql> AS
    csql> BEGIN
    csql>     return;
    csql>     dbms_output.put_line('Hello world');
    csql> END;

    ERROR: In line 5, column 5
    Stored procedure compile error: unreachable statement

```

bagus-kim commented on ko/pl/pl.rst:
```diff
@@ -107,15 +138,31 @@ SQL 구문 중에 다음에 해당하는 것들을 PL/CSQL 실행문으로 직
 
 SELECT 문은 실행문으로 사용될 뿐만 아니라 :ref:`커서 <cursor_decl>`\를 선언할 때나 :ref:`OPEN-FOR <cursor_manipulation>` 문에도 사용된다.
```
```suggestion
Static SQL은 String 변수에 담지 않고 코드 사이에 직접 기술한 SQL문을 의미한다. SELECT 문은 실행문으로 사용될 뿐만 아니라 :ref:`커서 <cursor_decl>`\를 선언할 때나 :ref:`OPEN-FOR <cursor_manipulation>` 문에도 사용된다.
```
Static SQL에 대한 정의를 추가했으면 합니다.제안 한 문구 또는 "Static SQL은 String 변수에 담지 않고 코드 사이에 직접 기술한 SQL문을 말합니다. "

hyunikn replied:
이 절의 맨 첫 부분에 "SQL 구문 중에 다음에 해당하는 것들을 PL/CSQL 실행문으로 직접 사용할 수 있으며, 그러한 경우를 Static SQL 문이라고 부른다." 라고 되어 있습니다.

bagus-kim commented on ko/pl/pl.rst:
```diff
@@ -190,15 +239,13 @@ Dynamic SQL은 주로 다음 두 가지 경우에 필요하다.
 
-식별자, 예약어, 주석, 리터럴을 작성할 때 Static/Dynamic SQL 안에서는
-`SQL의 작성 규칙 <https://www.cubrid.org/manual/ko/11.2/sql/syntax.html>`_\을 따른다.
+식별자, 예약어, 주석, 리터럴을 작성할 때 Static/Dynamic SQL 안에서는 :ref:`SQL의 작성 규칙 <lexical_rules>`\을 따른다.
```
```suggestion
Dynamic SQL은 String 변수에 담아서 기술하는 SQL문을 의미한다. 식별자, 예약어, 주석, 리터럴을 작성할 때 Static/Dynamic SQL 안에서는 :ref:`SQL의 작성 규칙 <lexical_rules>`\을 따른다.
```
여기서도 Dynamic SQL에 대한 설명필요하다 생각합니다.(

hyunikn replied:
작성 규칙 절 바로 위에 Dynamic SQL 절이 있으며, 절의 맨 첫 문장에 
"Dynamic SQL은 실행 시간에 SQL 구문에 해당하는 문자열을 만들어 [EXECUTE IMMEDIATE](http://192.168.2.208/ko/pl/pl.html#exec-imme) 문으로 실행하는 방식이다." 
라고 기술되어 있습니다.

hyunikn replied:
그 쪽으로 reference 를 추가하겠습니다.

bagus-kim commented on ko/pl/pl.rst:
```diff
@@ -265,31 +314,26 @@ Static/Dynamic SQL 밖의 PL/CSQL 문에서 아래 표의 단어들을 변수, 
 
-Static/Dynamic SQL에서는 SQL에서 제공하는 모든
-`데이터 타입 <https://www.cubrid.org/manual/ko/11.2/sql/datatype_index.html>`_\을 쓸 수 있다.
+Static/Dynamic SQL에서는 SQL에서 제공하는 모든 :ref:`데이터 타입 <datatype_index>`\을 쓸 수 있다.
```
```suggestion
Static/Dynamic SQL에서는 SQL에서 제공하는 모든 :ref:`데이터 타입 <datatype_index>`\을 사용할 수 있다.
```

hyunikn replied:
네. 수정하겠습니다.

bagus-kim commented on ko/pl/pl.rst:
```diff
@@ -343,8 +392,8 @@ SQL에서 제공하는 데이터 타입 중 PL/CSQL에서 지원하는 것과 
-나중에 그 컬럼의 타입이 변경되어도 자동으로 저장 프로시저/함수의 동작에 반영되지는 않는다.
+<table>.<column>%TYPE은 CREATE PROCEDURE/FUNTION 문을 실행하는 시점에 지정된 테이블 컬럼의 타입을 나타내지만,
+나중에 그 컬럼의 타입이 변경되어도 자동으로 <table>.<column>%TYPE을 사용한 저장 프로시저/함수의 동작에 반영되지는 않는다.
 그러므로, %TYPE을 적용한 테이블 컬럼의 타입이 변경되었을 때에는 그 %TYPE을 사용한 저장 프로시저/함수에 대해서 모두
```
```suggestion
그러므로, %TYPE을 적용한 테이블 컬럼의 타입이 변경되었을 때는 그 %TYPE을 사용한 저장 프로시저/함수에 대해서 모두
```

hyunikn replied:
네. 수정하겠습니다.

bagus-kim commented on ko/pl/pl.rst:
```diff
@@ -343,8 +392,8 @@ SQL에서 제공하는 데이터 타입 중 PL/CSQL에서 지원하는 것과 
+<table>.<column>%TYPE은 CREATE PROCEDURE/FUNTION 문을 실행하는 시점에 지정된 테이블 컬럼의 타입을 나타내지만,
+나중에 그 컬럼의 타입이 변경되어도 자동으로 <table>.<column>%TYPE을 사용한 저장 프로시저/함수의 동작에 반영되지는 않는다.
 그러므로, %TYPE을 적용한 테이블 컬럼의 타입이 변경되었을 때에는 그 %TYPE을 사용한 저장 프로시저/함수에 대해서 모두
 ALTER PROCEDURE/FUNCTION <name> REBUILD 문을 실행해서 재컴파일 해주어야 한다.
```
```suggestion
ALTER PROCEDURE/FUNCTION <name> REBUILD 문을 실행해서 재컴파일해 주어야 한다.
```

hyunikn replied:
네. 수정하겠습니다.

bagus-kim commented on ko/pl/pl.rst:
```diff
@@ -357,30 +406,166 @@ ALTER PROCEDURE/FUNCTION <name> REBUILD 문을 실행해서 재컴파일 해주
+
+.. code-block:: sql
+
+   CREATE TABLE tbl(a INT, b CHAR, c VARCHAR);
```
이건 추가 논의가 필요한 사항입니다만, 
매뉴얼에 기입되는 sample용 table및sql등은 demodb에 존재하는 것으로 변경이 필요합니다.

선별회의 논의 결과에 따라 진행 되겠지만, 새로이 매뉴얼 작성 시에는 demodb를 기준으로 작성하는 것이 좋지 않을까합니다.
필요하다면 demodb 스크립트를 수정 필요 하겠지만요~

hyunikn replied:
네. 큰 지장 없지 싶지만, 결정 기다리겠습니다.

bagus-kim replied:
12/16일 선별회의에서 
1. 매뉴얼예제인 경우, demodb기준으로 작성하자
2. 필요 시, demodb를 수정하자 
로 논의 되었습니다. 번거로우시겠지만 demodb를 수정하거나 아니면 demodb기준으로 샘플을 작성 요청 드립니다.

hyunikn replied:
네. 알겠습니다.

hyunikn replied:
이번 버전까지는 그대로 가고 guava 부터는 demodb 위주로 예제를 만들기로 했습니다.

bagus-kim commented on ko/pl/pl.rst:
```diff
@@ -497,6 +740,70 @@ Static/Dynamic SQL 문의 동작은 각종 `서버 설정 <https://www.cubrid.or
+:ref:`내부 프로시저/함수 선언 <local_routine_decl>`\과 :ref:`Block 실행문 <block_stmt>`\은
+자신만의 선언부와 실행부를 가지면서 중첩된 scope들을 이룬다.
+이 때 안쪽 scope 선언부에서 바깥에서 선언한 항목과 동일한 이름을 가진 다른 항목을 선언하면
+안쪽 scope에서 그 이름은 새로 정의한 항목을 가리키며 바깥쪽의 동일 이름은 가려진다.
```
```suggestion
안쪽 scope에서 그 이름은 새로 정의한 항목을 가리키며 바깥쪽의 동일 이름은 가려진다.이것을 "이름 가림" 이라한다.
```
"이름 가림" 이라는 용어가 영어에서 나왔다면 영어로 기입하는 것이 더 좋지 않을까요? 용어 선택이 어렵네요

hyunikn replied:
영어로는 'name hiding' 이라고 할 수 있습니다. 
(예: https://stackoverflow.com/questions/32946364/c-inheritance-and-name-hiding)

hyunikn replied:
영어 표현은 영문 버전에서만 써도 될 것 같습니다.

hyunikn committed:
. add examples
. apply review comments

hyunikn commented:
예제를 추가하였고, 주현님 리뷰 내용을 반영하였습니다. (5707f9d)
예제 추가까지가 이번 PR 에서 계획했던 최종 내용입니다.

hyunikn committed:
minor fix

hiclass commented on ko/pl/pl.rst:
```diff
@@ -357,30 +435,250 @@ ALTER PROCEDURE/FUNCTION <name> REBUILD 문을 실행해서 재컴파일 해주
+.. code-block:: sql
+
+    create table tblA(a INT, b CHAR, c VARCHAR);
+    create table tblB(a INT, b CHAR, c VARCHAR);        // tblA%ROWTYPE과 tblB%ROWTYPE은 동일 타입
```
```suggestion
    create table tblB(a INT, b CHAR, c VARCHAR);        -- tblA%ROWTYPE과 tblB%ROWTYPE은 동일 타입
```
// 도 주석표현방식이기는 하나 다양한 예제에서 SQL 방식의 주석으로 사용되었으므로 --로 변경하는 것을 권고합니다.

hyunikn replied:
네. 수정하겠습니다.

hiclass commented on ko/pl/pl.rst:
```diff
@@ -357,30 +435,250 @@ ALTER PROCEDURE/FUNCTION <name> REBUILD 문을 실행해서 재컴파일 해주
-
-* 서버 설정 파라미터 plus_as_concat 값에 상관없이 +가 숫자 덧셈과 문자열 병합 연산자로 쓰인다.
+* 나머지 연산자 %를 쓸 수 없다. 단, 동일한 의미의 MOD를 대신 쓸 수 있다.
+* 논리 연산자 &&, ||, ! 들을 쓸 수 없다. 단, 각각 동일한 의미의 AND, OR, NOT을 대신 쓸 수 있다.
```
```suggestion
* 논리 연산자 &&, ||, ! 들을 쓸 수 없다. 단, 각각 동일한 의미의 AND, OR, NOT을 대신 쓸 수 있다.
```
실제 동작이 제한되는 것인지 확인 부탁 드립니다.

hyunikn replied:
대체 기호가 있는 경우 구현 기간을 줄이기 위해 넣지 않았습니다. 
CUBRID SQL 스펙에 비해 부족한 부분이 있으나 차후 버전에서 추가하려고 합니다.

hyunikn replied:
실제 동작이 제한되는 것을 예제와 코드에서 확인했습니다.

hiclass commented on ko/pl/pl.rst:
```diff
@@ -357,30 +435,250 @@ ALTER PROCEDURE/FUNCTION <name> REBUILD 문을 실행해서 재컴파일 해주
-  + 특히, ||는 서버 설정 파라미터 pipes_as_concat 값에 상관없이 논리합 연산자로 쓰이지 않는다.
-
-* 서버 설정 파라미터 plus_as_concat 값에 상관없이 +가 숫자 덧셈과 문자열 병합 연산자로 쓰인다.
+* 나머지 연산자 %를 쓸 수 없다. 단, 동일한 의미의 MOD를 대신 쓸 수 있다.
```
```suggestion
* 나머지 연산자 %를 쓸 수 없다. 단, 동일한 의미의 MOD를 대신 쓸 수 있다.
```
실제 동작이 제한되는 것인지 확인 부탁 드립니다.

hyunikn replied:
대체 기호가 있는 경우 구현 기간을 줄이기 위해 넣지 않았습니다.
CUBRID SQL 스펙에 비해 부족한 부분이 있으나 차후 버전에서 추가하려고 합니다.

hyunikn replied:
%의 경우는 따라오는 단어가 found, isopen, notfound, rowcount, rowtype, type 인 경우 다른 의미로 해석되기 때문에 %를 나머지 연산자로 사용하려면 구현 방법을 위한 특별한 고민이 필요합니다. 현재 구현에서는 다음 예제에서처럼 변수명이 하필 found 인 경우 원래 의도와 다르게 해석되어 에러를 일으킵니다. 
```
csql> create or replace procedure poo as
csql>     found int := 1;
csql>     i int := 5;
csql> begin
csql>     i := i % found;
csql> end;

In line 5, column 10,

ERROR: In line 5, column 10
Stored procedure compile error: cursor attributes may not be read from I which is neither a cursor nor a cursor reference

```

hyunikn replied:
실제 동작이 제한되는 것을 예제와 코드에서 확인했습니다.

bagus-kim committed:
[CUBRIDMAN-222]  매뉴얼 설명 수정  및 예제 수정 (#553)

* RAND()범위 (2 31미만 -> 2^31미만)으로 내용 수정

* 가독성을 위한 sql 예제문 수정

* 국문 rst오류 수정 / 영문 예제수정

* http://www.cubrid.com:8888/browse/RND-2290
예약어 목록에 'ACCESS'추가

* 국문파일 누락되어 commit
http://www.cubrid.com:8888/browse/RND-2290 예약어 목록에 'ACCESS'추가

* http://www.cubrid.com:8888/browse/RND-2316
1. 시스템 설정 목록 수정
2. ;set에 명령어 link처리

* [CUBRIDMAN-222] http://www.cubrid.com:8888/browse/RND-2291
block_ddl_statement가 yes인경우 통계 정보 갱신 에러가 발생할 있다는 내용 추가

* csql> 및 불필요한 문자열 삭제

* **block_ddl_statement** 에서 no --> **no**로 수정

* http://www.cubrid.com:8888/browse/RND-2234

단어수정 : 디렉토리 -> 디렉터리

* server_stmt.rst에서 :ref:`cci_connect_with_url` link오류 수정

* http://www.cubrid.com:8888/browse/RND-2336
IS NULL연산자 예제 수정 : CREATE INDEX idx ON t (a) WHERE (not a) IS NULL; -> CREATE INDEX idx ON ta (a) WHERE (not ta.a<>0 ) IS NULL;

* http://www.cubrid.com:8888/browse/RND-2309
예제문 수정 : ALTER TABLE t_def a a int; -> ALTER TABLE t_def MODIFY a int;

hyunikn committed:
replace // with --

airnet73 committed:
[CUBRIDMAN-252] Extends the loadBalance Property of JDBC (#557)

http://jira.cubrid.org/browse/CUBRIDMAN-252

The usage of the loadBalance property of the connection url has been expanded, and a manual has been written for the added settings.

shparkcubrid committed:
[CUBRIDMAN-253] manual for HASH LIST SCAN for Hierarchical Queries (#559)

http://jira.cubrid.org/browse/CUBRIDMAN-253

Shows that a hash list scan is performed when executing a hierarchical query with join tables.

hyunikn commented:
리뷰 시작한지 2주가 흘렀습니다. 
리뷰 마무리 부탁 드립니다.

mhoh3963 commented on ko/csql.rst:
```diff
@@ -1364,6 +1403,41 @@ OPT LEVEL의 상세한 내용은 :ref:`viewing-query-plan`\ 를 참고한다.
+
+이 값을 ON으로 설정하면 서버의 DBMS_OUTPUT 버퍼에 저장된 메시지를 출력한다. 기본값은 OFF이다.
+DBMS_OUTPUT 버퍼는 주로 PL/CSQL 저장 프로시저/함수에서 DBMS_OUTPUT.put_line() 호출을 통해 쌓인 메시지들을 저장하고 있다.
+DBMS_OUTPUT 메시지는 CSQL이 실행한 SQL 문의 결과 출력에 뒤이어 '<DBMS_OUTPUT>' 이라는 표시 다음에 출력된다.
```
```suggestion
DBMS_OUTPUT 메시지들은 CSQL이 실행한 SQL 문의 결과 출력 후에 '<DBMS_OUTPUT>'  표시와 함께 출력된다.
```

mhoh3963 commented on ko/csql.rst:
```diff
@@ -1364,6 +1403,41 @@ OPT LEVEL의 상세한 내용은 :ref:`viewing-query-plan`\ 를 참고한다.
+이 값을 ON으로 설정하면 서버의 DBMS_OUTPUT 버퍼에 저장된 메시지를 출력한다. 기본값은 OFF이다.
+DBMS_OUTPUT 버퍼는 주로 PL/CSQL 저장 프로시저/함수에서 DBMS_OUTPUT.put_line() 호출을 통해 쌓인 메시지들을 저장하고 있다.
+DBMS_OUTPUT 메시지는 CSQL이 실행한 SQL 문의 결과 출력에 뒤이어 '<DBMS_OUTPUT>' 이라는 표시 다음에 출력된다.
+특히, 실행 순서 상 DBMS_OUTPUT.put_line() 호출이 먼저이고 에러 발생이 나중이라도 실행 결과인 에러메시지가 먼저 출력되고
```
```suggestion
저장프로시스에서 오류 발생시 DBMS_OUTPUT.put_line()으로 출력한 메세지들은 실행 순서와 관계없이 에러메시지 출력 후 출력되므로 주의가 필요하다.
```

mhoh3963 commented on ko/csql.rst:
```diff
@@ -1364,6 +1403,41 @@ OPT LEVEL의 상세한 내용은 :ref:`viewing-query-plan`\ 를 참고한다.
+DBMS_OUTPUT 버퍼는 주로 PL/CSQL 저장 프로시저/함수에서 DBMS_OUTPUT.put_line() 호출을 통해 쌓인 메시지들을 저장하고 있다.
+DBMS_OUTPUT 메시지는 CSQL이 실행한 SQL 문의 결과 출력에 뒤이어 '<DBMS_OUTPUT>' 이라는 표시 다음에 출력된다.
+특히, 실행 순서 상 DBMS_OUTPUT.put_line() 호출이 먼저이고 에러 발생이 나중이라도 실행 결과인 에러메시지가 먼저 출력되고
+DBMS_OUTPUT 메시지 출력이 나중이니 주의하자. ::
```
```suggestion
```

youngjinj committed:
[CUBRIDMAN-241] Support HASH JOIN with hint required (#556)

http://jira.cubrid.org/browse/CUBRIDMAN-241
http://jira.cubrid.org/browse/CBRD-25382

Add descriptions of Nested Loop Join, Sort Merge Join, and Hash Join to Query Optimization section

hyunikn committed:
Update ko/csql.rst - minor rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/csql.rst - minor rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/csql.rst - minor rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

mhoh3963 commented on ko/csql.rst:
```diff
@@ -341,7 +341,42 @@ CSQL 시작 옵션
+    csql> create table table_c([;exit] int);
+    ...
+
+여기서 주의할 한 가지 예외가 있다. 이들 영역 안일지라도 CSQL이 대화형 모드이고 명령행을 세미콜론으로 시작하면
```
```suggestion
예외적으로 대화형 모드 CSQL에서는 명령행의 첫문자가 세미콜론인 경우 리터럴, 주석, 식별자 처리중이여도 해당 문자열은 세션 명령어로 인식한다. ::
```

mhoh3963 commented on ko/csql.rst:
```diff
@@ -341,7 +341,42 @@ CSQL 시작 옵션
+    ...
+
+여기서 주의할 한 가지 예외가 있다. 이들 영역 안일지라도 CSQL이 대화형 모드이고 명령행을 세미콜론으로 시작하면
+세션 명령어로 인식한다는 것이다. ::
```
```suggestion
```

mhoh3963 commented on ko/csql.rst:
```diff
@@ -341,7 +341,42 @@ CSQL 시작 옵션
+    csql> ;exit    <-- CSQL 종료
+    ...
+
+이 성질은 CSQL을 종료한다거나 (:ref:`;exit <scmd_exit>`),
```
```suggestion
이 예외는 CSQL을 종료한다거나 (:ref:`;exit <scmd_exit>`),
```

beyondykk9 committed:
[CUBRIDMAN-223] Extending result-cache to uncorrelated subqueries (#519)

http://jira.cubrid.org/browse/CUBRIDMAN-223

mhoh3963 commented on ko/csql.rst:
```diff
@@ -581,9 +620,9 @@ CSQL 인터프리터에서 작업 중인 데이터베이스 이름 및 호스트
+**파라미터 값 설정(;SET)**
 
-특정 파라미터의 값을 설정하기 위해서는 **;Set** 세션 명령어를 사용한다. 동적 변경이 가능한 파라미터만 값을 변경할 수 있으며, 서버 파라미터는 DBA 권한이 있어야만 값을 변경할 수 있다. 동적 변경이 가능한 파라미터 목록은 :ref:`broker-configuration` 를 참고한다. ::
+특정 파라미터의 값을 설정하기 위해서는 **;set** 세션 명령어를 사용한다. 동적 변경이 가능한 파라미터만 값을 변경할 수 있으며, 서버 파라미터는 DBA 권한이 있어야만 값을 변경할 수 있다. 동적 변경이 가능한 파라미터 목록은 :ref:`broker-configuration` 를 참고한다. ::
```
해당 내용은 develop branch에 머지시 확인 후 머지가 필요합니다.

hyunikn replied:
"DBA 권한이 있어야만" 이 확인할 부분일까요?

mhoh3963 replied:
";SET" command 문자열 수정 부분입니다. 대현님이 develop에 관련된 부분을 수정했을꺼예요.

hyunikn replied:
대현님께 확인했습니다. (https://github.com/CUBRID/cubrid-manual/pull/550) 
대문자 표시는 축약표시를 위한 것이고, 여기서는 그 정보를 표시하지 않아도 될 것 같아 소문자로 바꾸었습니다.
CSQL 의 help 메시지 밖에서도 축약표시 정보가 필요할까요.

hyunikn replied:
develop 브랜치에 merge 할 때 다른 부분과 일관성을 고려하여 조정하겠습니다.

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -29,6 +31,24 @@ PL/CSQL은 저장 프로시저나 저장 함수를 생성하는데 사용된다.
+저장 프로시저/함수는 :ref:`큐브리드 내장 함수 <operators-and-functions>`\와 동일한 이름을 가질 수 없다.
+동일한 이름으로 선언하면 컴파일 과정에서 (CREATE 문 실행 과정에서) 에러가 발생한다.
+
+body 내부의 실행문들 중에 실행 중 도달할 수 없는 것이 있을 때는 컴파일 과정에서 에러가 발생한다.
```
```suggestion
body 내부의 실행문들 중 수행시 도달할 수 없는 실햄문이 있는 경우 컴파일 과정에서 에러를 발생한다.
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -89,6 +109,25 @@ PL/CSQL은 저장 프로시저나 저장 함수를 생성하는데 사용된다.
             RETURN -1;
     END;
 
+CREATE PROCEDURE/FUNCTION 문을 실행하면 저장 프로시저/함수의 문법과 실행의미에 관련된 각종 규칙들을 검사한다.
```
```suggestion
CREATE PROCEDURE/FUNCTION 문을 실행하면 저장 프로시저/함수의 문법과 실행 의미에 관련된 각종 규칙들을 검사한다.
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -89,6 +109,25 @@ PL/CSQL은 저장 프로시저나 저장 함수를 생성하는데 사용된다.
     END;
 
+CREATE PROCEDURE/FUNCTION 문을 실행하면 저장 프로시저/함수의 문법과 실행의미에 관련된 각종 규칙들을 검사한다.
+검사에서 오류가 발견되면 오류의 위치와 원인을 설명하는 메시지를 클라이언트 응용 프로그램으로 전달한다.
```
```suggestion
검사에서 오류가 발견되면 발생된 위치와 원인을 설명하는 오류 메세지를 출력한다.
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -102,20 +141,56 @@ SQL 구문 중에 다음에 해당하는 것들을 PL/CSQL 실행문으로 직
+        ...
+    END;
+
+SELECT 문이 INTO 절을 포함한 경우 조회 결과는 한 건 그리고 단 한 건의 결과 레코드를 가져야 한다.
```
```suggestion
 INTO절을 포함안 SELECT 문의 조회 결과는 한 건의 결과 레코드여야 하며, 두건 이상의 결과를 가져오는 경우는 TOO_MANY_ROWS Exception이 발생한다.
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -102,20 +141,56 @@ SQL 구문 중에 다음에 해당하는 것들을 PL/CSQL 실행문으로 직
+    END;
+
+SELECT 문이 INTO 절을 포함한 경우 조회 결과는 한 건 그리고 단 한 건의 결과 레코드를 가져야 한다.
+결과가 없을 때는 NO_DATA_FOUND Exception이 발생하고 결과가 두 건 이상일 때는 TOO_MANY_ROWS Exception이 발생한다.
```
```suggestion
결과가 없을 경우에는 NO_DATA_FOUND Exception이 발생한다.
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -357,30 +435,250 @@ ALTER PROCEDURE/FUNCTION <name> REBUILD 문을 실행해서 재컴파일 해주
+* s와 t의 필드 개수가 같다.
+* 각각의 필드 순번 i에 대해서, s와 t의 i번째 필드들의 타입을 S\ :sub:`i`\와 T\ :sub:`i`\라고 할 때, S\ :sub:`i`\에서 T\ :sub:`i`\로 대입 가능하다.
+
+레코드 변수 사이에 대입이 가능하기 위해서 같은 순번의 필드 이름들끼리 같을 필요는 없다.
```
```suggestion
레코드 변수 사이에 대입이 가능하기 위해서 같은 순번의 필드명이 동일할 필요는 없다.
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -357,30 +435,250 @@ ALTER PROCEDURE/FUNCTION <name> REBUILD 문을 실행해서 재컴파일 해주
+        r1 := r3;   -- Error: TIME에서 DATETIME으로 대입 불가 (형변환 불가)
+    END;
+
+%ROWTYPE은 내부 프로시저/함수의 인자 타입과 리턴 타입으로 쓸 수 있다.
```
```suggestion
%ROWTYPE은 내부 프로시저/함수의 인자 타입과 리턴 타입으로 사용할 수 있다.
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -357,30 +435,250 @@ ALTER PROCEDURE/FUNCTION <name> REBUILD 문을 실행해서 재컴파일 해주
+    END;
+
+%ROWTYPE은 내부 프로시저/함수의 인자 타입과 리턴 타입으로 쓸 수 있다.
+그러나, 저장 프로시저/함수의 인자 타입과 리턴 타입으로는 쓸 수 없다. SQL 문에서 레코드 타입을 지원하지 않기 때문이다.
```
```suggestion
그러나, SQL문에서 레코드 타입을 지원하지 않기 때문에 저장 프로시저/함수의 인자 타입과 리턴 타입으로는 사용할 수 없다. 
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -357,30 +435,250 @@ ALTER PROCEDURE/FUNCTION <name> REBUILD 문을 실행해서 재컴파일 해주
+        ...
+    END;
+
+Static/Dynamic SQL SELECT 문과 FETCH 문의 INTO 절에 레코드 변수를 쓸 수 있다.
```
```suggestion
Static/Dynamic SQL SELECT 문과 FETCH 문의 INTO 절에 레코드 변수를 사용할 수 있으며, 레코드 변수를 사용할 경우 다른 변수에 함께 사용할 수는 없다.
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -357,30 +435,250 @@ ALTER PROCEDURE/FUNCTION <name> REBUILD 문을 실행해서 재컴파일 해주
+    END;
+
+Static/Dynamic SQL SELECT 문과 FETCH 문의 INTO 절에 레코드 변수를 쓸 수 있다.
+단, 이 때 INTO 절 안에 다른 변수를 함께 쓸 수 없다.
```
```suggestion
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -357,30 +435,250 @@ ALTER PROCEDURE/FUNCTION <name> REBUILD 문을 실행해서 재컴파일 해주
+
+Static/Dynamic SQL SELECT 문과 FETCH 문의 INTO 절에 레코드 변수를 쓸 수 있다.
+단, 이 때 INTO 절 안에 다른 변수를 함께 쓸 수 없다.
+그리고, 조회 결과를 이루는 컬럼들과 레코드 필드들의 개수가 같아야 한다.
```
```suggestion
그리고, 조회 결과와 레코드 변수의 컬럼명은 동일할 필요는 없지만, 조회 컬럼 개수와 레코드 변수의 컬럼 개수는 동일해야 하며, 조회 컬럼 타입과 레코드 변수의 컬럼 타입이 동일하거나 호환 가능해야 한다.
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -357,30 +435,250 @@ ALTER PROCEDURE/FUNCTION <name> REBUILD 문을 실행해서 재컴파일 해주
+Static/Dynamic SQL SELECT 문과 FETCH 문의 INTO 절에 레코드 변수를 쓸 수 있다.
+단, 이 때 INTO 절 안에 다른 변수를 함께 쓸 수 없다.
+그리고, 조회 결과를 이루는 컬럼들과 레코드 필드들의 개수가 같아야 한다.
+이 때, 같은 순번의 컬럼과 필드끼리 이름이 같을 필요는 없지만 컬럼 타입으로부터 필드 타입으로 대입 가능해야 한다.
```
```suggestion
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -357,30 +435,250 @@ ALTER PROCEDURE/FUNCTION <name> REBUILD 문을 실행해서 재컴파일 해주
+   -- Fetch
+   FETCH c INTO part;
+
+Static SQL INSERT/REPLACE 문의 VALUES 절에 레코드 변수를 쓸 수 있다.
```
```suggestion
Static SQL INSERT/REPLACE 문의 VALUES 절에 레코드 변수를 사용할 수 있으며, 레코드 변수를 사용할 경우 다른 변수와 함께 사용할 수 없다.
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -357,30 +435,250 @@ ALTER PROCEDURE/FUNCTION <name> REBUILD 문을 실행해서 재컴파일 해주
+   FETCH c INTO part;
+
+Static SQL INSERT/REPLACE 문의 VALUES 절에 레코드 변수를 쓸 수 있다.
+단, 이 때 VALUES 절 안에 다른 변수를 함께 쓸 수 없다.
```
```suggestion
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -357,30 +435,250 @@ ALTER PROCEDURE/FUNCTION <name> REBUILD 문을 실행해서 재컴파일 해주
+
+Static SQL INSERT/REPLACE 문의 VALUES 절에 레코드 변수를 쓸 수 있다.
+단, 이 때 VALUES 절 안에 다른 변수를 함께 쓸 수 없다.
+그리고, 대입되는 테이블 컬럼들과 레코드 필드들의 개수가 같아야 한다.
```
```suggestion
그리고, 대입 컬럼명과 레코드 변수의 컬럼명은 동일할 필요는 없지만, 대입 컬럼 개수와 레코드 변수의 컬럼 개수는 동일해야 하며, 대입 컬럼 타입과 레코드 변수의 컬럼 타입이 동일하거나 호환 가능해야 한다.
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -357,30 +435,250 @@ ALTER PROCEDURE/FUNCTION <name> REBUILD 문을 실행해서 재컴파일 해주
+Static SQL INSERT/REPLACE 문의 VALUES 절에 레코드 변수를 쓸 수 있다.
+단, 이 때 VALUES 절 안에 다른 변수를 함께 쓸 수 없다.
+그리고, 대입되는 테이블 컬럼들과 레코드 필드들의 개수가 같아야 한다.
+이 때, 같은 순번의 컬럼과 필드끼리 이름이 같을 필요는 없지만 필드 타입으로부터 컬럼 타입으로 대입 가능해야 한다.
```
```suggestion
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -357,30 +435,250 @@ ALTER PROCEDURE/FUNCTION <name> REBUILD 문을 실행해서 재컴파일 해주
+
+그리고, 일반적으로 정밀도와 스케일이 생략된 NUMERIC은 NUMERIC(15, 0)을 의미하지만
+예외적으로 인자 타입과 리턴 타입 자리에서는 임의의 정밀도와 스케일을 허용하는 것으로 동작한다
+(단, 정밀도는 1 이상 38 이하. 스케일은 0 이상 정밀도 이하 범위 안에서).
```
```suggestion
(단, 정밀도 범위 : 1 <= 정밀도 <= 38  스케일 범위 : 0 <= 스케일 <= 정밀도).
```

mhoh3963 replied:
```suggestion
(단, 정밀도는 1 이상 38 이하,  스케일은 0 이상 정밀도 이하 범위).
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -357,30 +435,250 @@ ALTER PROCEDURE/FUNCTION <name> REBUILD 문을 실행해서 재컴파일 해주
+그리고, 일반적으로 정밀도와 스케일이 생략된 NUMERIC은 NUMERIC(15, 0)을 의미하지만
+예외적으로 인자 타입과 리턴 타입 자리에서는 임의의 정밀도와 스케일을 허용하는 것으로 동작한다
+(단, 정밀도는 1 이상 38 이하. 스케일은 0 이상 정밀도 이하 범위 안에서).
+또한, CHAR와 VARCHAR도 인자 타입과 리턴 타입 자리에서는 다른 곳에서처럼 CHAR(1)과 VARCHAR(1073741823)를 나타내는 것이
```
```suggestion
또한, CHAR와 VARCHAR도 인자 타입과 리턴 타입 자리에서는 기본 스케일 값인 CHAR(1)과 VARCHAR(1073741823)를 나타내는 것이
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -357,30 +435,250 @@ ALTER PROCEDURE/FUNCTION <name> REBUILD 문을 실행해서 재컴파일 해주
+(단, 정밀도는 1 이상 38 이하. 스케일은 0 이상 정밀도 이하 범위 안에서).
+또한, CHAR와 VARCHAR도 인자 타입과 리턴 타입 자리에서는 다른 곳에서처럼 CHAR(1)과 VARCHAR(1073741823)를 나타내는 것이
+아니라 임의의 길이를 갖는 문자열을 허용하는 것으로 동작한다
+(단, CHAR 길이는 2048 이하. VARCHAR의 길이는 1073741823 이하 범위 안에서).
```
```suggestion
(단, CHAR 길이는 2048 이하, VARCHAR의 길이는 1073741823 이하 범위).
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -357,30 +435,250 @@ ALTER PROCEDURE/FUNCTION <name> REBUILD 문을 실행해서 재컴파일 해주
+    SELECT test_any_length('abcd');     -- 결과: 'abcd'
+
+인자 타입과 리턴 타입을 :ref:`%TYPE <percent_type>`\을 사용해서 지정했을 때에도 참조되는 원래 타입의
+정밀도, 스케일, 길이 지정은 무시되고 대신 임의의 정밀도, 스케일, 길이를 허용하는 것으로 동작한다.
```
```suggestion
정밀도, 스케일 및 길이 지정은 무시되고 대신 임의의 정밀도, 스케일, 길이를 허용하는 것으로 동작한다.
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -459,19 +767,95 @@ PL/CSQL은 다른 많은 프로그래밍 언어와 마찬가지로 Exception 핸
     END;
 
-서버 설정 적용 예외
+발생한 Exception이 마지막까지 WHEN ... THEN ... 절로 처리되지 않은 경우에는
```
```suggestion
발생한 Exception에 대한 처리를 명시적 (WHEN ... THEN ... 절)으로 처리하지 않은 경우에는
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -459,19 +767,95 @@ PL/CSQL은 다른 많은 프로그래밍 언어와 마찬가지로 Exception 핸
 
-서버 설정 적용 예외
+발생한 Exception이 마지막까지 WHEN ... THEN ... 절로 처리되지 않은 경우에는
+코드상에서의 Exception 발생 위치와 에러메시지가 DBMS에 접속한 클라이언트 응용 프로그램에 전달된다.
```
```suggestion
코드상의 Exception 발생 위치와 에러메시지가 출력된다.
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -459,19 +767,95 @@ PL/CSQL은 다른 많은 프로그래밍 언어와 마찬가지로 Exception 핸
-Static/Dynamic SQL 문의 동작은 각종 `서버 설정 <https://www.cubrid.org/manual/ko/11.2/admin/config.html#id2>`_\의 영향을 받는다.
+Static/Dynamic SQL 문의 동작은 :ref:`서버 설정 파라미터 <system_config>` 전체의 영향을 동일하게 받는다.
+
+Static/Dynamic SQL 밖에서 PL/CSQL 문은 오직 다음 4개 서버 설정 파라미터만이 유효하다.
```
```suggestion
Static/Dynamic SQL 제외한 PL/CSQL 문에서는 다음 4개 서버 설정 파라미터만이 유효하다.
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -459,19 +767,95 @@ PL/CSQL은 다른 많은 프로그래밍 언어와 마찬가지로 Exception 핸
+
+이들 설정의 자세한 의미는 :ref:`서버 설정 파라미터 <system_config>`\를 참조할 수 있다.
+
+위 4개 외 다른 설정은 Static/Dynamic SQL 밖의 PL/CSQL 문에서 유효하지 않다. 특히,
```
```suggestion
위 4개 외 다른 설정은 Static/Dynamic SQL 제외한 PL/CSQL 문에서 유효하지 않다. 특히,
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -497,6 +881,70 @@ Static/Dynamic SQL 문의 동작은 각종 `서버 설정 <https://www.cubrid.or
+    END;
+
+이러한 "이름 가림"은 다른 종류의 항목(상수, 프로시저/함수 인자, Exception, 커서, 내부 프로시저/함수)들에
+대해서도 마찬가지로 적용된다.
```
```suggestion
대해서도 동일하게 적용된다.
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -574,13 +988,16 @@ NOT NULL 조건이 지정된 경우에는 반드시 NULL이 아닌 초기값이
 * *builtin_type*: :ref:`데이터 타입 <types>` 절에서 설명한 시스템 제공 타입
 
 상수 선언에는 필수적으로 값 지정이 포함되어야 한다.
+NOT NULL 조건이 지정된 경우에 이 값은 NULL이 아니어야 한다.
```
```suggestion
NOT NULL 조건이 지정된 경우, 이 값은 NULL이 아니어야 한다.
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -953,15 +1384,55 @@ SYS_REFCURSOR 변수에 연결하고 그 SELECT 문의 결과를 조회해 오
+RAISE_APPLICATION_ERROR는 원하는 :ref:`코드와 에러메시지 <sqlcode>`\로 :ref:`Exception <exception>`\을
+일으키고자 할 때 사용한다.
+RAISE_APPLICATION_ERROR의 사용 형태는 Built-in 프로시저 호출처럼 보이지만 내부적으로는 PL/CSQL 실행문이다.
+첫번째 인자로 주는 코드는 1000보다 큰 INTEGER 값을 가져야 한다. 아니면, VALUE_ERROR Exception이 발생한다.
```
```suggestion
첫번째 인자로 주는 코드는 1000보다 큰 INTEGER 값을 가져야 한다. 1000 이하의 값인 경우 시스템 예약값에 의해 해당 Exception 대신 VALUE_ERROR Exception이 발생한다.
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -953,15 +1384,55 @@ SYS_REFCURSOR 변수에 연결하고 그 SELECT 문의 결과를 조회해 오
+일으키고자 할 때 사용한다.
+RAISE_APPLICATION_ERROR의 사용 형태는 Built-in 프로시저 호출처럼 보이지만 내부적으로는 PL/CSQL 실행문이다.
+첫번째 인자로 주는 코드는 1000보다 큰 INTEGER 값을 가져야 한다. 아니면, VALUE_ERROR Exception이 발생한다.
+1000 이하의 값은 시스템을 위해 예약되어 있기 때문이다.
```
```suggestion
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -953,15 +1384,55 @@ SYS_REFCURSOR 변수에 연결하고 그 SELECT 문의 결과를 조회해 오
+값들은 대응되는 변수나 OUT 인자에 대입 가능한 타입을 가져야 한다.
+
+SQL 문 실행 중에 에러가 나면 SQL_ERROR Exception이 발생한다.
+INTO 절을 포함한 경우 SELECT 문의 조회 결과는 한 건 그리고 단 한 건의 결과 레코드를 가져야 한다.
```
```suggestion
INTO 절을 포함한 경우 SELECT 문의 조회 결과는  단 한 건의 결과 레코드를 가져야 한다.
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -973,12 +1444,12 @@ INTO 절을 써서 SELECT 문의 조회 결과를 프로그램의 변수나 OUT
-  포함된 표현식의 갯수는 일치해야 한다.
-* *using_clause*: *dynamic_sql*\을 실행할 때 문자열의 ? 자리에 채워질 값들을 지정한다.
-* *into_clause*: *dynamic_sql*\이 SELECT문을 나타내는 경우에 조회 결과를 담을 변수나 OUT 파라미터를 지정한다.
+  SQL 구문 중간중간 값을 필요로 하는 자리에 ?(물음표)를 대신 쓸 수 있으며 이러한 ?의 개수와 *using_clause*\에
```
```suggestion
  SQL 구문 내부에서 값을 필요로 하는 자리에 ?(물음표)를 대신 쓸 수 있으며 사용한 ?의 개수와 *using_clause*\에
```

mhoh3963 commented on ko/pl/pl.rst:
```diff
@@ -1323,25 +1818,24 @@ PL/CSQL의 표현식의 종류는 다음 문법으로 요약할 수 있다.
 
-* 커서와 연관되지 않은 SELECT 문의 경우 반드시 INTO 절을 포함하게 되고 조회 결과는 1개이어야 한다. 따라서, 이 SELECT 문이 정상적으로 수행되었을 때 SQL%ROWCOUNT의 값은 1이다. 조회 결과 크기가 0이거나 1을 초과해서 실행시간 에러가 발생했을 때에는 SQL%ROWCOUNT의 값은 정의되지 않는다.
-* INSERT, UPDATE, DELETE, MERGE, REPLACE, TRUNCATE 문의 경우 영향 받은 레코드 갯수가 된다.
+* 커서와 연관되지 않은 SELECT 문의 경우 반드시 INTO 절을 포함하게 되고 조회 결과는 1개이어야 한다. 따라서, 이 SELECT 문이 정상적으로 수행되었을 때 SQL%ROWCOUNT의 값은 1이다. 조회 결과 크기가 0이거나 1을 초과해서 실행시간 에러가 발생했을 때는 SQL%ROWCOUNT의 값은 정의되지 않는다.
```
```suggestion
* 커서와 연관되지 않은 SELECT 문의 경우 반드시 INTO 절을 사용해야 하며 조회 결과는 1개이어야 한다. 따라서, 이 SELECT 문이 정상적으로 수행되었을 때 SQL%ROWCOUNT의 값은 1이다. 조회 결과 크기가 0이거나 1을 초과해서 실행시간 에러가 발생했을 때는 SQL%ROWCOUNT의 값은 정의되지 않는다.
```

mhoh3963 commented:
@hyunikn 
수정된 내용에 대해서는 1차 리뷰했습니다. PL/CSQL의 전체 내용은 develop 머지시 다시 한번 리뷰하겠습니다.

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/csql.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/csql.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/csql.rst

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn committed:
Update ko/pl/pl.rst - rephrase

Co-authored-by: Mike MyungHwan Oh <mhoh@cubrid.com>

hyunikn commented:
> @hyunikn 수정된 내용에 대해서는 1차 리뷰했습니다. PL/CSQL의 전체 내용은 develop 머지시 다시 한번 리뷰하겠습니다.

감사합니다. 한 가지 질문을 남겼고 모두 반영했습니다.

hyunikn committed:
rephrase

hyunikn committed:
minor typo

beyondykk9 committed:
[CUBRIDMAN-240] max_query_per_tran 설정 값 관련 설명 추가 (#544)

http://jira.cubrid.org/browse/CUBRIDMAN-240

hyunikn commented:
시간 관계 상 이번 주 금요일까지 리뷰 내용으로 갈음하고 머지하려고 합니다. 
리뷰 마무리 부탁 드립니다.

shparkcubrid committed:
[CUBRIDMAN-254] Manual for COUNT_ONLY optimization and Enhancing group_concat() (#560)

http://jira.cubrid.org/browse/CUBRIDMAN-254

http://jira.cubrid.org/browse/CBRD-24248
http://jira.cubrid.org/browse/CBRD-24850

Add COUNT_ONLY optimization on covering indexes and Enhancing group_concat().

hyunikn commented:
시간 관계 상 지금까지의 리뷰까지 반영하여 마무리 하겠습니다.

hyunikn committed:
Merge branch 'develop' into feature/plcsql

hyunikn committed:
Merge branch 'feature/plcsql' into phase1-pl-features
